kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/AdapterGenerator.kt

```diff
@@ -185,10 +199,9 @@ internal class AdapterGenerator(
           result.addStatement("%N = %N.fromJson(%N)",
               property.localName, nameAllocator[property.delegateKey], readerParam)
         } else {
-          result.addStatement("%N = %N.fromJson(%N) ?: throw·%T(%P)",
+          result.addStatement("%N = %N.fromJson(%N) ?: throw·jsonDataException(%S, %S, %S, %N)",
```

Comment:
What if we extract a helper method in code gen instead to generate these?

```kotlin
fun jsonDataException(description: String, identifier: String, condition: String, readerParam: ParameterSpec): CodeBlock {
  val message = CodeBlock.builder()
      .beginControlFlow("buildString")
      .addStatement("append(%S)", description)
      .addStatement("append(%S)", identifier)
      .addStatement("append(%S)", condition)
      .addStatement("append(%N.path)", readerParam)
      .endControlFlow()
      .build()
  return CodeBlock.of("%T(%L)", JsonDataException::class, message)
}
```

Or with your logic

```kotlin
fun jsonDataException(description: String, identifier: String, condition: String, readerParam: ParameterSpec): CodeBlock {
  return CodeBlock.of("%T(%T(%S).append(%S).append(%S).append(%N.path).toString())", JsonDataException::class, StringBuilder::class, description, identifier, condition, jsonReader)
}
```

Reply:
Yes, i like this idea. Tried it and seems to be working. R8 also outlines most of the code (it outlines the first part of the string concatenation `StringBuilder(a).append(b).append(c)` into one method, and it outlines the second part `sb.append(reader.getPath()).toString()` into another method. But all in all it still deduplicates the data, and the resulting dex file is even a little smaller then with the helper method.